### ISSUE TITLE
configure workflowr project to NOT push to remote

### ISSUE THREAD

**Issue by rgayler** [workflowr/workflowr#141]
I work on confidential projects, where the thought of accidentally pushing to a GitHub remote is horrifying.

As a safety measure I have defined local versions of wflow_git_remote() and wflow_git_push() that call stop("Don't do that!). However, that doesn't prevent some function calling workflowr::wflow_git_push().

Would it be possible to add a local_only=TRUE argument to wflow_start() that takes steps to ensure that workflowr never tries to have dealings with a remote repo?

**Comment by pcarbo** [comment]
> The thought of accidentally pushing to a GitHub remote is horrifying.

@rgayler If this is a concern, it seems that the simplest option would be to remove the GitHub remote, and at that point `wflow_git_push` should no longer work. For more details, run `git remote  --help`.

**Comment by rgayler** [comment]
I never even created a GitHub remote in the first place, but I will eventually be handing over the project directory to people who are even more clueless than I am about workflowr and git - so I redefined wflow_git_remote() to avoid accidents via that route (although, obviously, that doesn't stop somebody using git to push directly to a remote).

Given that lots of research projects have confidentiality constraints I thought it might provide a degree of comfort for workflowr users if there was some project-level configuration that could be set to indicate the project is confidential and prevent pushing to a remote repo.

(Maybe a [git pre-push hook](https://blog.ghost.org/prevent-master-push/ could be used to prevent git from pushing to a remote.)

**Comment by jdblischak** [comment]
> I work on confidential projects, where the thought of accidentally pushing to a GitHub remote is horrifying.

@rgayler Could you please describe your use case some more so that we can better understand how to support your use of workflowr? For example, how are you sharing the workflowr website with your collaborators? Do you have a private web server or are you sending files over email? Did you use a remote Git hosting service of any kind, or do you plan to only use Git locally? How do you plan to share the code/data with your collaborators?

> Would it be possible to add a local_only=TRUE argument to wflow_start() that takes steps to ensure that workflowr never tries to have dealings with a remote repo?

Since this is the first request for this feature, I'm not ready to add an argument for it. For more complicated Git behavior, I think it is best to directly use Git (see below).

> (Maybe a [git pre-push hook](https://blog.ghost.org/prevent-master-push/ could be used to prevent git from pushing to a remote.)

I think this is the right solution. I saved the file below as `.git/hooks/pre-push`, made it executable with `chmod +x .git/hooks/pre-push`, and confirmed that it prevents pushing to a remote repository.

```
#!/bin/bash

echo "This is a confidential project. Do not push the files to a remote server"
echo "For details, contact Ross Gayler via ..."
exit 1
```

Now this hook file is not transferred using the traditional Git commands (`push`, `pull`, `clone`), but I think that is ok for your use case since you don't seem to be using those commands. If you share the workflowr project directly with your collaborators (e.g. using a file sharing service or emailing a tarball), the pre-push hook will prevent them from pushing to GitHub or any other Git hosting website.

**Comment by rgayler** [comment]
**Use case:**
My current use case is that I am working on a project for a third party. The project is contractually constrained to be confidential - I can only share project materials with the third party.
During the course of the project I need to repeatedly share results (but not code) for feedback.
At the end of the project the entire project directory is sent to the third party so they have the ability to reproduce and modify the analyses if they wish.

I like the [Dat/Beakersuggestion](https://jdblischak.github.io/workflowr/articles/wflow-05-faq.html#can-i-share-workflowr-websites-directly-with-collaborators-in-a-secure-fashion) in the FAQ,
but the third-parties I work with tend to have a hard time getting anything non-standard into their IT environments.

I use Git locally in the project directory to track changes to the code base. I protect against data loss with encrypted backup.

I currently share the knitted results with the third party by sending an encrypted archive of the `docs` directory.

At the end of the project I will send an encrypted archive of the whole project directory to the third party.

Generalising my use case to more likely scientific scenarios, I think the key features are:

1. The project contains or may generate information that must be kept confidential. This would be fairly common in any field that deals with human data, because there is a risk of having personally identifiable information. Even if you expect the analyses to only generate unidentifiable aggregate figures there is a risk that someone will accidentally slip some PII into the output text, e.g. when identifying an aberrant case that needs to be followed up ("John Smith has negative age").

2. The project has collaborators who don't need to modify the code, e.g. subject-matter experts. They only need to review the outputs.

> Since this is the first request for this feature, I'm not ready to add an argument for it.

I think what I was after was firstly a declarative statement of intent to keep the project local. E.g. it could be a line in `_workflowr.yml` or the existence of a file named `CONFIDENTIAL.txt` in the project directory. Then given the declaration that the project is not to be published (unintentionally) any mechanisms to prevent accidental publication via Git are essentially implementation detail.

> I think [a Git pre-push hook] is the right solution.

So the suggestion was that something like that might be built in to `workflowr` so that users who know almost nothing about Git (e.g. me) might have more confidence in using `workflowr` for confidential projects.

**Comment by jdblischak** [comment]
@rgayler Thanks for the explanation. It is very insightful to learn the various ways that workflowr is being used.

> At the end of the project I will send an encrypted archive of the whole project directory to the third party.

Great. This means that the pre-push hook is a good solution for this situation.

> The project contains or may generate information that must be kept confidential.

I agree this is a common situation. My problem trying to generalize it is that disabling Git is not necessarily required in this situation. For example, if the organization uses [GitHub Enterprise](https://github.com/enterprise), then the analysts can push to the remote repository even though the data is confidential.

> it could be a line in _workflowr.yml or the existence of a file named CONFIDENTIAL.txt in the project directory.

I think adding such a file is best done manually. It's not as common as say a README.md file.

> So the suggestion was that something like that might be built in to workflowr so that users who know almost nothing about Git (e.g. me) might have more confidence in using workflowr for confidential projects.

How about an argument to `wflow_start()` named something like `disable_remote`? When set to `TRUE`, it would create the file `.git/hooks/pre-push` and make it executable. I'd propose the file contain something like the following (suggestions welcome):

```
#!/bin/bash

# This hook prevents pushing to a remote repository, which is ideal for a
# project containing confidential data. It was created by wflow_start() with the
# argument disable_remote = TRUE. If you decide you want to be able to push this
# repository, delete this file.

echo "This is a confidential project. Do not push the files to a remote server"
exit 1
```

**Comment by rgayler** [comment]
> if the organization uses GitHub Enterprise, then the analysts can push to the remote repository even though the data is confidential.

1. So the core concept is "not to be pushed to a public repo" (which may follow from confidentiality).
2. The feature is aimed at users who are not technically sophisticated (or have access to technical sophistication).

> wflow_start() named something like disable_remote? When set to TRUE, it would create the file .git/hooks/pre-push

That sounds great. The existence of the argument makes it clear that the possibility exists and absolves the technically naive user of having to come up with a technical solution. I just redefined `wflow_git_remote()` and `wflow_git_push()` to give an error message, but the pre-push hook prevents pushing by any mechanism. The mechanism doesn't have to be absolutely watertight (a malign actor can always find a workaround) it just needs to give a degree of well-founded confidence that the user is unlikely to accidentally expose the project.

**Comment by jdblischak** [comment]
@rgayler OK. I've made a note to add the argument `disable_remote` to `wflow_start()`.

**Comment by rgayler** [comment]
Thanks John

**Comment by jdblischak** [comment]
@rgayler What operating systems do you and your colleagues use? Getting the pre-push hook solution  to work for Windows as well appears like it will be difficult.

**Comment by jdblischak** [comment]
For my future reference: [StackOverflow discussion of Git hooks on Windows](https://stackoverflow.com/questions/18277429/executing-git-hooks-on-windows)

**Comment by jdblischak** [comment]
@rgayler This feature is now available in [workflowr 1.2.0](https://github.com/jdblischak/workflowr/releases/tag/v1.2.0). You can install it from the GitHub master branch or wait for it to be built and synced across the CRAN mirrors.
